<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Изменение-типов-данных." data-toc-modified-id="Изменение-типов-данных.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Изменение типов данных.</a></span></li><li><span><a href="#Удаление-дубликатов." data-toc-modified-id="Удаление-дубликатов.-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Удаление дубликатов.</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Категоризация-дохода." data-toc-modified-id="Категоризация-дохода.-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Категоризация дохода.</a></span></li><li><span><a href="#Категоризация-целей-кредита." data-toc-modified-id="Категоризация-целей-кредита.-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Категоризация целей кредита.</a></span></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-9.1"><span class="toc-item-num">9.1&nbsp;&nbsp;</span>Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-1:-Из-полученных-данных-видно,-что-зависимость-между-количеством-детей-и-возвратом-кредита-отсутствует." data-toc-modified-id="Вывод-1:-Из-полученных-данных-видно,-что-зависимость-между-количеством-детей-и-возвратом-кредита-отсутствует.-9.2"><span class="toc-item-num">9.2&nbsp;&nbsp;</span>Вывод 1: Из полученных данных видно, что зависимость между количеством детей и возвратом кредита отсутствует.</a></span></li><li><span><a href="#Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-9.3"><span class="toc-item-num">9.3&nbsp;&nbsp;</span>Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-2:-Из-полученных-данных-видно,-что-лучше-всего-возвращают-кредит-вдовцы/вдовы,-за-ними-идут-разведенные,-следующие-те,-кто-состоит-в-браке,-еще-хуже-те,-кто-состоит-в-гражданском-браке,-ну-и-на-последнем-месте---не-женатые-/-не-замужние." data-toc-modified-id="Вывод-2:-Из-полученных-данных-видно,-что-лучше-всего-возвращают-кредит-вдовцы/вдовы,-за-ними-идут-разведенные,-следующие-те,-кто-состоит-в-браке,-еще-хуже-те,-кто-состоит-в-гражданском-браке,-ну-и-на-последнем-месте---не-женатые-/-не-замужние.-9.4"><span class="toc-item-num">9.4&nbsp;&nbsp;</span>Вывод 2: Из полученных данных видно, что лучше всего возвращают кредит вдовцы/вдовы, за ними идут разведенные, следующие те, кто состоит в браке, еще хуже те, кто состоит в гражданском браке, ну и на последнем месте - не женатые / не замужние.</a></span></li><li><span><a href="#Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-9.5"><span class="toc-item-num">9.5&nbsp;&nbsp;</span>Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-3-:-Из-полученных-данных-видно,-что-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок-отсутствует.-Чаще-остальных-кредит-возвращают-заемщики-с-уровнем-дохода-D-(30001–50000),-реже-остальных-возвращают-заемщики-с-уровнем-дохода-E-(0-30000)." data-toc-modified-id="Вывод-3-:-Из-полученных-данных-видно,-что-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок-отсутствует.-Чаще-остальных-кредит-возвращают-заемщики-с-уровнем-дохода-D-(30001–50000),-реже-остальных-возвращают-заемщики-с-уровнем-дохода-E-(0-30000).-9.6"><span class="toc-item-num">9.6&nbsp;&nbsp;</span>Вывод 3 : Из полученных данных видно, что зависимость между уровнем дохода и возвратом кредита в срок отсутствует. Чаще остальных кредит возвращают заемщики с уровнем дохода D (30001–50000), реже остальных возвращают заемщики с уровнем дохода E (0-30000).</a></span></li><li><span><a href="#Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-9.7"><span class="toc-item-num">9.7&nbsp;&nbsp;</span>Вопрос 4: Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Вывод-4:-Из-полученных-данных-видно,-что-реже-остальных-просрачивают-возврат-кредита-заемщики,-взявшие-кредит-на-недвижимость---7,46%,-чаще-возвращают-те,-кто-брал-кредит-на-проведение-свадьбы---8%,-с-такой-же-частотой-просрачивают-кредит-заемщики,-взявшие-кредит-на-образование---8%,-на-последнем-месте-стоят-те,-кто-взял-кредит-на-автомобиль---9,36%." data-toc-modified-id="Вывод-4:-Из-полученных-данных-видно,-что-реже-остальных-просрачивают-возврат-кредита-заемщики,-взявшие-кредит-на-недвижимость---7,46%,-чаще-возвращают-те,-кто-брал-кредит-на-проведение-свадьбы---8%,-с-такой-же-частотой-просрачивают-кредит-заемщики,-взявшие-кредит-на-образование---8%,-на-последнем-месте-стоят-те,-кто-взял-кредит-на-автомобиль---9,36%.-9.8"><span class="toc-item-num">9.8&nbsp;&nbsp;</span>Вывод 4: Из полученных данных видно, что реже остальных просрачивают возврат кредита заемщики, взявшие кредит на недвижимость - 7,46%, чаще возвращают те, кто брал кредит на проведение свадьбы - 8%, с такой же частотой просрачивают кредит заемщики, взявшие кредит на образование - 8%, на последнем месте стоят те, кто взял кредит на автомобиль - 9,36%.</a></span></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

# Исследование надежности заемщиков.

В данном проекте нашим заказчиком является банк, предоставляющий кредиты клентам. 
В проекте требуется оценить влияние различных факторов на возврат кредита в срок.
Результаты будут использованы для модели кредитного скорринга.

## Обзор данных

Импортируем библиотеки

In [1]:
import pandas as pd
import os

Загрузим данные

In [2]:
pth1 = 'data.csv'
pth2 = 'https://code.s3.yandex.net/datasets/data.csv'
    
if os.path.exists(pth1):
    data = pd.read_csv(pth1)
elif os.path.exists(pth2):
    data = pd.read_csv(pth2)
else:
    print('Something is wrong')

Посмотрим первые 5 элементов таблицы для общего представления о данных

In [3]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Посмотрим общую информацию о таблице

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Посмотрим на статистическую информацию о данных

In [5]:
data.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

Данные представляют из себя таблицу из 21525 строк (заемщики банка) и 12 столбцов (данные о заемщиках).  

В полученных данных у нас следующие столбцы:  
children — количество детей в семье  
days_employed — общий трудовой стаж в днях  
dob_years — возраст клиента в годах    
education — уровень образования клиента  
education_id — идентификатор уровня образования  
family_status — семейное положение  
family_status_id — идентификатор семейного положения  
gender — пол клиента  
income_type — тип занятости  
debt — имел ли задолженность по возврату кредитов  
total_income — ежемесячный доход  
purpose — цель получения кредита.

В данных имеются пропуски в столбцах days_employed и total_income - потребуется выбрать какими значениями их заполнить, 

есть отрицательные значения в столбцах children и days_employed - потребуется избавиться от отрицательных значений, 

в столбце dob_years есть нулевые значения - потребуется выбрать на какое значение их поменять.

## Заполнение пропусков

Из данных видно, что пропуски есть только в столбцах 'days_employed' и 'total_income' и их количество совпадает. 
Скорее всего оба столбца не заполнены в одних и тех же строчках. Проверим это.

In [6]:
data[data.isnull().any(1)] 

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

Количество строк таблицы с пропущенными элементами совпадает с количеством пропусков, значит данные отсуствуют в одних и тех же строках.
Возможно две причины отсутсвия данных:
- данные о работе были пропущены при заполнении таблицы;
- заемщик на данный момент не имеет постоянной работы.

In [7]:
percenf_of_na = data['days_employed'].isna().mean()
percenf_of_na # доля количества строк с пропусками к количеству строк всего в процентах

0.10099883855981417

Перед заполнением пропущенных данных необходимо избавиться от аномалий в этих столбцах.

В столбце days_employed присутствуют отрицательные значения - заменим их на значения по модулю.  
Возможно, что при заполнении таблицы сотрудник перепутал ячейки для заполнения "работает с" и "работает по"

In [8]:
data['days_employed'] = data['days_employed'].abs() 

Переведем столбец со стажем из дней в года и переименум его.

In [9]:
data['days_employed'] = data['days_employed'] / 365
data = data.rename(columns={'days_employed' : 'years_employed'})

In [10]:
data.describe()

children  years_employed     dob_years  education_id  \
count  21525.000000    19351.000000  21525.000000  21525.000000   
mean       0.538908      183.328024     43.293380      0.817236   
std        1.381587      380.906522     12.574584      0.548138   
min       -1.000000        0.066141      0.000000      0.000000   
25%        0.000000        2.539751     33.000000      1.000000   
50%        0.000000        6.011563     42.000000      1.000000   
75%        1.000000       15.172281     53.000000      1.000000   
max       20.000000     1100.699727     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

Максимальное значение стажа работы - 1100, что не может быть правдой.
Определим реальный максимальный стаж.

In [11]:
data[data['years_employed'] < 500].sort_values(by='years_employed', ascending=False).head(5)

children  years_employed  dob_years education  education_id  \
16335         1       50.380685         61   среднее             1   
4299          0       48.261817         61   среднее             1   
7329          0       45.461569         60    высшее             0   
17838         0       44.560821         59   среднее             1   
16825         0       44.163528         64   среднее             1   

         family_status  family_status_id gender income_type  debt  \
16335  женат / замужем                 0      F   сотрудник     0   
4299   женат / замужем                 0      F   компаньон     0   
7329   женат / замужем                 0      F   сотрудник     0   
17838  женат / замужем                 0      F   сотрудник     0   
16825  женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
16335  186178.934089      операции с недвижимостью  
4299   122560.741753                 покупка жилья  
7329   124697.846781  заняться высшим образованием  
17838   51238.967133         на покупку автомобиля  
16825   91527.685995    покупка жилой недвижимости

Из полученных данных видно, что реальный максимальный стаж не более 51 года.  
Проверим как связаны столбцы years_employed и income_type у клиентов, максимальный стаж которых более 51 года.

In [12]:
data[data['years_employed'] > 51]['income_type'].value_counts()

пенсионер      3443
безработный       2
Name: income_type, dtype: int64

3443 из них - пенсионеры.  
Посмотрим статистические данные о пенсионерах.

In [13]:
data[data['income_type'] == 'пенсионер'].describe()

children  years_employed    dob_years  education_id  \
count  3856.000000     3443.000000  3856.000000   3856.000000   
mean      0.132002     1000.009565    59.063019      0.913900   
std       1.014271       57.724948     7.618526      0.510185   
min      -1.000000      900.626632     0.000000      0.000000   
25%       0.000000      949.724236    56.000000      1.000000   
50%       0.000000     1000.584401    60.000000      1.000000   
75%       0.000000     1049.949033    64.000000      1.000000   
max      20.000000     1100.699727    74.000000      4.000000   

       family_status_id         debt   total_income  
count       3856.000000  3856.000000    3443.000000  
mean           0.985737     0.056017  137127.465690  
std            1.315202     0.229984   80246.953231  
min            0.000000     0.000000   20667.263793  
25%            0.000000     0.000000   82881.443465  
50%            0.000000     0.000000  118514.486412  
75%            2.000000     0.000000  169700.433010  
max            4.000000     1.000000  735103.270167

Минимальное количество лет стажа - 900, что не может быть правдой.  
Средний возраст пенсионеров в наших данных - 60 лет, в связи с этим можем заменить общий трудовой стаж у всех пенсионеров на медианное значение стажа среди клинтов возрастом старше 55 лет, для которых у нас есть "корректные" данные по трудовому стажу.

In [14]:
data.loc[(data['income_type'] == 'пенсионер') & (data['years_employed'] > 51), 'years_employed'] = \
    data.loc[(data['years_employed'] < 51) & (data['dob_years'] > 55), 'years_employed'].median()

Проверим что получилось.

In [15]:
data[data['income_type'] == 'пенсионер'].describe()

children  years_employed    dob_years  education_id  \
count  3856.000000    3.443000e+03  3856.000000   3856.000000   
mean      0.132002    6.588070e+00    59.063019      0.913900   
std       1.014271    2.584975e-13     7.618526      0.510185   
min      -1.000000    6.588070e+00     0.000000      0.000000   
25%       0.000000    6.588070e+00    56.000000      1.000000   
50%       0.000000    6.588070e+00    60.000000      1.000000   
75%       0.000000    6.588070e+00    64.000000      1.000000   
max      20.000000    6.588070e+00    74.000000      4.000000   

       family_status_id         debt   total_income  
count       3856.000000  3856.000000    3443.000000  
mean           0.985737     0.056017  137127.465690  
std            1.315202     0.229984   80246.953231  
min            0.000000     0.000000   20667.263793  
25%            0.000000     0.000000   82881.443465  
50%            0.000000     0.000000  118514.486412  
75%            2.000000     0.000000  169700.433010  
max            4.000000     1.000000  735103.270167

Посмотрим на оставшихся клиентов с аномальными значениями стажа.

In [16]:
data[data['years_employed'] > 51]

children  years_employed  dob_years education  education_id  \
3133          1      924.724567         31   среднее             1   
14798         0     1083.021476         45    Высшее             0   

          family_status  family_status_id gender  income_type  debt  \
3133    женат / замужем                 0      M  безработный     1   
14798  гражданский брак                 1      F  безработный     0   

        total_income                  purpose  
3133    59956.991984  покупка жилья для сдачи  
14798  202722.511368             ремонт жилью

Таких осталось всего два клиента, поэтому данные для этих клиентов заменим на меданное значение стажа всех остальных клиентов

In [17]:
data.loc[data['years_employed'] > 51, 'years_employed'] = data['years_employed'].median()

In [18]:
data.describe()

children  years_employed     dob_years  education_id  \
count  21525.000000    19351.000000  21525.000000  21525.000000   
mean       0.538908        6.471742     43.293380      0.817236   
std        1.381587        5.723761     12.574584      0.548138   
min       -1.000000        0.066141      0.000000      0.000000   
25%        0.000000        2.539751     33.000000      1.000000   
50%        0.000000        6.011563     42.000000      1.000000   
75%        1.000000        7.527188     53.000000      1.000000   
max       20.000000       50.380685     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

In [19]:
data[data.isnull().any(1)].describe()

children  years_employed    dob_years  education_id  \
count  2174.000000             0.0  2174.000000   2174.000000   
mean      0.552438             NaN    43.632015      0.800828   
std       1.469356             NaN    12.531481      0.530157   
min      -1.000000             NaN     0.000000      0.000000   
25%       0.000000             NaN    34.000000      0.250000   
50%       0.000000             NaN    43.000000      1.000000   
75%       1.000000             NaN    54.000000      1.000000   
max      20.000000             NaN    73.000000      3.000000   

       family_status_id         debt  total_income  
count       2174.000000  2174.000000           0.0  
mean           0.975161     0.078197           NaN  
std            1.418220     0.268543           NaN  
min            0.000000     0.000000           NaN  
25%            0.000000     0.000000           NaN  
50%            0.000000     0.000000           NaN  
75%            1.000000     0.000000           NaN  
max            4.000000     1.000000           NaN

После того как избавились от аномалий в столбцах, где присутствуют пропуски, займемся заполнением пропусков.

На уровень дохода влияет уровень образования. Заполним пропущенные значения в total_income медианными значениями по каждой категории education_id.

In [20]:
for id in data['education_id'].unique():
    
    data.loc[data['education_id'] == id, 'total_income'] = \
        data.loc[data['education_id'] == id, 'total_income']\
        .fillna(data.loc[data['education_id'] == id, 'total_income'].median())

In [21]:
data[data.isnull().any(1)].describe()

children  years_employed    dob_years  education_id  \
count  2174.000000             0.0  2174.000000   2174.000000   
mean      0.552438             NaN    43.632015      0.800828   
std       1.469356             NaN    12.531481      0.530157   
min      -1.000000             NaN     0.000000      0.000000   
25%       0.000000             NaN    34.000000      0.250000   
50%       0.000000             NaN    43.000000      1.000000   
75%       1.000000             NaN    54.000000      1.000000   
max      20.000000             NaN    73.000000      3.000000   

       family_status_id         debt   total_income  
count       2174.000000  2174.000000    2174.000000  
mean           0.975161     0.078197  146766.496306  
std            1.418220     0.268543   17133.460261  
min            0.000000     0.000000  117137.352825  
25%            0.000000     0.000000  136478.643244  
50%            0.000000     0.000000  136478.643244  
75%            1.000000     0.000000  171534.463802  
max            4.000000     1.000000  175340.818855

На стаж работы влияет возраст клиента, поэтому разобьем клиентов на три группы и заполним пропуски медианными значениями по группам.  
Группы:
- до 34 лет
- от 35 до 54 лет
- старше 55 лет  

In [22]:
data.loc[data['dob_years'] <= 34, 'years_employed'] = \
    data.loc[data['dob_years'] <= 34, 'years_employed'].\
    fillna(data.loc[data['dob_years'] <= 34, 'years_employed'].median()) #заполняем пропуски для первой группы
data.loc[(data['dob_years'] > 34) & (data['dob_years'] <= 54), 'years_employed'] = \
    data.loc[(data['dob_years'] > 34) & (data['dob_years'] <= 54), 'years_employed'].\
    fillna(data.loc[(data['dob_years'] > 34) & (data['dob_years'] <= 54), 'years_employed'].median()) #заполняем пропуски для второй группы
data.loc[data['dob_years'] > 54, 'years_employed'] = \
    data.loc[data['dob_years'] > 54, 'years_employed'].\
    fillna(data.loc[data['dob_years'] > 54, 'years_employed'].median()) #заполняем пропуски для третьей группы

In [23]:
data.isna().sum()

children            0
years_employed      0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Пропуски в данных отсутствуют.

## Проверка данных на аномалии и исправления.

Cнова посмотрим на первые строки таблицы, чтобы понять где искать аномалии.

In [24]:
data.head()  

children  years_employed  dob_years education  education_id  \
0         1       23.116912         42    высшее             0   
1         1       11.026860         36   среднее             1   
2         0       15.406637         33   Среднее             1   
3         3       11.300677         32   среднее             1   
4         0        6.588070         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Проверяем уникальные значения количества детей

In [25]:
data['children'].unique() 

array([ 1,  0,  3,  2, -1,  4, 20,  5])

"-1" - явно ошибка исправим на "1"

In [26]:
data['children'] = data['children'].abs()

Обработаем аномалии в возрасте клиентов, у которых возраст указан 0 лет.

In [27]:
data[data['dob_years'] == 0].describe()

children  years_employed  dob_years  education_id  family_status_id  \
count  101.000000      101.000000      101.0    101.000000        101.000000   
mean     0.673267        5.847487        0.0      0.673267          1.237624   
std      2.098137        5.188776        0.0      0.512033          1.524129   
min      0.000000        0.298540        0.0      0.000000          0.000000   
25%      0.000000        2.790480        0.0      0.000000          0.000000   
50%      0.000000        4.414687        0.0      1.000000          1.000000   
75%      1.000000        6.588070        0.0      1.000000          3.000000   
max     20.000000       29.285618        0.0      2.000000          4.000000   

             debt   total_income  
count  101.000000     101.000000  
mean     0.079208  157327.989397  
std      0.271410   70855.379784  
min      0.000000   34974.450366  
25%      0.000000  102621.701671  
50%      0.000000  136537.132625  
75%      0.000000  201852.430096  
max      1.000000  386373.636958

Так как стаж и возраст скорее всего связаны, то разобьем этих клиентов на 3 группы по стажу и заполним медианными значениями по группам:
- до 2.8 лет
- от 2.8 до 6.6 лет
- от 6.6 лет

In [28]:
data.loc[(data['years_employed'] <= 2.8) & (data['dob_years'] == 0), 'dob_years'] = \
    data.loc[(data['years_employed'] <= 2.8)& (data['dob_years'] != 0), 'dob_years'].median() #заполняем пропуски для первой группы
data.loc[(data['years_employed'] > 2.8) & (data['years_employed'] <= 6.6) & (data['dob_years'] == 0), 'dob_years'] = \
    data.loc[(data['years_employed'] > 2.8) & (data['years_employed'] <= 6.6)  & (data['dob_years'] != 0), 'dob_years'].median() #заполняем пропуски для второй группы
data.loc[(data['years_employed'] > 6.6) & (data['dob_years'] == 0), 'dob_years'] = \
    data.loc[(data['years_employed'] > 6.6) & (data['dob_years'] != 0), 'dob_years'].median() #заполняем пропуски для третьей группы

In [29]:
data.describe()

children  years_employed     dob_years  education_id  \
count  21525.000000    21525.000000  21525.000000  21525.000000   
mean       0.543275        6.362139     43.499001      0.817236   
std        1.379876        5.453124     12.222997      0.548138   
min        0.000000        0.066141     19.000000      0.000000   
25%        0.000000        2.809885     34.000000      1.000000   
50%        0.000000        5.987423     43.000000      1.000000   
75%        1.000000        6.899093     53.000000      1.000000   
max       20.000000       50.380685     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  2.152500e+04  
mean           0.972544      0.080883  1.653361e+05  
std            1.420324      0.272661  9.798248e+04  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.077982e+05  
50%            0.000000      0.000000  1.409601e+05  
75%            1.000000      0.000000  1.955436e+05  
max            4.000000      1.000000  2.265604e+06

В данных присутствуют клиенты, у которых 20 детей, посмотрим статистические данные по этим клиентам.

In [30]:
data[data['children'] == 20].describe()

children  years_employed  dob_years  education_id  family_status_id  \
count      76.0       76.000000  76.000000     76.000000         76.000000   
mean       20.0        5.945745  42.447368      0.842105          0.815789   
std         0.0        5.688632  11.061519      0.433671          1.363432   
min        20.0        0.331507  21.000000      0.000000          0.000000   
25%        20.0        2.101947  34.000000      1.000000          0.000000   
50%        20.0        5.766228  42.000000      1.000000          0.000000   
75%        20.0        6.674300  50.000000      1.000000          1.000000   
max        20.0       32.704579  69.000000      2.000000          4.000000   

            debt   total_income  
count  76.000000      76.000000  
mean    0.105263  166436.441711  
std     0.308931   76368.794728  
min     0.000000   53971.576721  
25%     0.000000  120986.388504  
50%     0.000000  142073.192806  
75%     0.000000  203490.401462  
max     1.000000  441721.334145

Таких клиентов всего 76. Так как это не очень большое количество от общего количества клиентов и ввиду трудности нахождения связи с другими столбцами, то поменяем значения для всех таких клиентов на медианное значение количества детей.

In [31]:
data.loc[data['children'] == 20, 'children'] = data['children'].median()

In [32]:
data.describe()

children  years_employed     dob_years  education_id  \
count  21525.000000    21525.000000  21525.000000  21525.000000   
mean       0.472660        6.362139     43.499001      0.817236   
std        0.750616        5.453124     12.222997      0.548138   
min        0.000000        0.066141     19.000000      0.000000   
25%        0.000000        2.809885     34.000000      1.000000   
50%        0.000000        5.987423     43.000000      1.000000   
75%        1.000000        6.899093     53.000000      1.000000   
max        5.000000       50.380685     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  2.152500e+04  
mean           0.972544      0.080883  1.653361e+05  
std            1.420324      0.272661  9.798248e+04  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.077982e+05  
50%            0.000000      0.000000  1.409601e+05  
75%            1.000000      0.000000  1.955436e+05  
max            4.000000      1.000000  2.265604e+06

Все аномалии устранены.

## Изменение типов данных.

Изменим тип данных 'total_income' на целочисленный,  
изменим тип данных 'years_employed' на тип данных с плавающей точной,  
изменим тип данных количества детей на целочисленное

In [33]:
data['total_income'] = data['total_income'].astype('int32') 
data['years_employed'] = data['years_employed'].astype('float32') 
data['children'] = data['children'].astype('int8') 
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int8   
 1   years_employed    21525 non-null  float32
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  int32  
 11  purpose           21525 non-null  object 
dtypes: float32(1), int32(1), int64(4), int8(1), object(5)
memory usage: 1.7+ MB


## Удаление дубликатов.

Посмотрим уникальные значения в столбце 'education'

In [34]:
data['education'].unique()

array(['высшее', 'среднее', 'Среднее', 'СРЕДНЕЕ', 'ВЫСШЕЕ',
       'неоконченное высшее', 'начальное', 'Высшее',
       'НЕОКОНЧЕННОЕ ВЫСШЕЕ', 'Неоконченное высшее', 'НАЧАЛЬНОЕ',
       'Начальное', 'Ученая степень', 'УЧЕНАЯ СТЕПЕНЬ', 'ученая степень'],
      dtype=object)

Все люди заполняют таблицы по разному, кто-то пишет все в нижнем регистре, кто-то пишет с закглавной буквы, а кто-то заполняет все с включеным CAPS LOCK

Сделаем все записи в нижнем регистре

In [35]:
data['education'] = data['education'].str.lower() 

Посмотрим уникальные значения в столбце 'family_status'

In [36]:
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

Сделаем все записи в нижнем регистре

In [37]:
data['family_status'] = data['family_status'].str.lower()

Посмотрим уникальные значения в столбце 'gender'. XNA предположу, что не определился или трансгендер

In [38]:
data['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

Посмотрим уникальные значения в столбце 'gender'

In [39]:
data['income_type'].unique()

array(['сотрудник', 'пенсионер', 'компаньон', 'госслужащий',
       'безработный', 'предприниматель', 'студент', 'в декрете'],
      dtype=object)

Посмотрим уникальные значения в столбце 'purpose'

In [40]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Посмотрим количество явных дубликатов после наших действий

In [41]:
data.duplicated().sum()

71

Избавимся от явных дубликатов

In [42]:
data = data.drop_duplicates().reset_index(drop=True)

## Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Создадим словари 'education_dict' и 'family_status'

In [43]:
education_dict = data[['education', 'education_id']].drop_duplicates()
family_status_dict = data[['family_status', 'family_status_id']].drop_duplicates()

Удалим столбцы 'education' и 'family_status'

In [44]:
data = data.drop(['education', 'family_status'], axis=1)

## Категоризация дохода.

Создадим функцию 'total_income_group', разделяющую заемщиков на группы по уровню дохода

In [45]:
def total_income_group(total_income):
    if total_income <= 30000:
        return 'E'
    elif total_income <= 50000:
        return 'D'
    elif total_income <= 200000:
        return 'C'
    elif total_income <= 1000000:
        return 'B'
    else:
        return 'A'

Создадим столбец 'total_income_category' с группами по уровню дохода

In [46]:
data['total_income_category'] = data['total_income'].apply(total_income_group)

In [47]:
data.head()

children  years_employed  dob_years  education_id  family_status_id gender  \
0         1       23.116913         42             0                 0      F   
1         1       11.026859         36             1                 0      F   
2         0       15.406637         33             1                 0      M   
3         3       11.300677         32             1                 0      M   
4         0        6.588070         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

## Категоризация целей кредита.

Создадим функцию 'purpose_group', разделяющую заемщиков на группы по целям кредита

In [48]:
def purpose_group(purpose):
    if 'авто' in purpose:
        return 'операции с автомобилем'
    elif 'недвиж' in purpose or 'жиль' in purpose:
        return 'операции с недвижимостью'
    elif 'свадьб' in purpose:
        return 'проведение свадьбы'
    else:
        return 'получение образования'

Создадим столбец 'purpose_category' с группами по целям кредита

In [49]:
data['purpose_category'] = data['purpose'].apply(purpose_group)

In [50]:
data['purpose_category'].value_counts()

операции с недвижимостью    10811
операции с автомобилем       4306
получение образования        4013
проведение свадьбы           2324
Name: purpose_category, dtype: int64

In [51]:
data.head()

children  years_employed  dob_years  education_id  family_status_id gender  \
0         1       23.116913         42             0                 0      F   
1         1       11.026859         36             1                 0      F   
2         0       15.406637         33             1                 0      M   
3         3       11.300677         32             1                 0      M   
4         0        6.588070         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

## Ответы на вопросы.

### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [52]:
data.pivot_table(index='children', 
                 values='debt', 
                 aggfunc=['count', 'sum', 'mean'])\
                    .style.format({('mean', 'debt'): '{:.2%}'})

### Вывод 1: Из полученных данных видно, что зависимость между количеством детей и возвратом кредита отсутствует.
Лучше всего возвращают кредит заемщики без детей (7,54% просрачивают кредит). Возможной причиной являются отсутствие больших ежемесячных трат (кружки, секции, питание детей), более легкое перераспределение своих доходов (ограничить расходы на детей гораздо сложнее).  
Количество клиентов, у которых 4 и 5 детей очень мало, поэтому считаю, что эти данные могут поменяться, если выборка будет больше.

### Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [53]:
data.pivot_table(index='family_status_id', values='debt', aggfunc=['count', 'sum', 'mean'])

count  sum      mean
                   debt debt      debt
family_status_id                      
0                 12339  931  0.075452
1                  4151  388  0.093471
2                   959   63  0.065693
3                  1195   85  0.071130
4                  2810  274  0.097509

### Вывод 2: Из полученных данных видно, что лучше всего возвращают кредит вдовцы/вдовы, за ними идут разведенные, следующие те, кто состоит в браке, еще хуже те, кто состоит в гражданском браке, ну и на последнем месте - не женатые / не замужние.
Предположительно вдовцы/вдовы и разведенные привыкли рассчитывать только на свои доходы, но при этом им могут помогать родственники с обеих сторон, заемщики, состоящие в гражданском браке ,возможно, при оформлении кредита рассчитывали на доход гражданского супруга/супруги, но после расстования дохода уже может не хватать, холостые, возможно, еще не научились оценивать свои возможности и ежемесячные выплаты по кредиту значительно снижают уровень комфорта, к которому они привыкли.


### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [54]:
data.pivot_table(index='total_income_category', values='debt', aggfunc=['count', 'sum', 'mean'])

count   sum      mean
                        debt  debt      debt
total_income_category                       
A                         25     2  0.080000
B                       5041   356  0.070621
C                      16016  1360  0.084915
D                        350    21  0.060000
E                         22     2  0.090909

### Вывод 3 : Из полученных данных видно, что зависимость между уровнем дохода и возвратом кредита в срок отсутствует. Чаще остальных кредит возвращают заемщики с уровнем дохода D (30001–50000), реже остальных возвращают заемщики с уровнем дохода E (0-30000).
Заемщики с не очень высоким уровнем дохода могут легче сменить работу в случае падения уровня дохода на текущем месте работы. Заемщикам с высоким уровнем дохода, в случае непредвиденных обстоятельств, треднее сменить место работы на работу со схожими доходами и сложнее перестроить расходы семьи, так как ближайшие родственники привыкли к текущему уровню дохода (дети учатся в хороших частных школах, посещают платные секции, учаться платно в университете).  

Однако, в крайних группах по уровню дохода (A и E) у нас мало данных - 25 и 22, соответственно, поэтому данные по этим группам для модели кредитного скоринга не годятся.

### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

In [55]:
data.pivot_table(index='purpose_category', values='debt', aggfunc=['count', 'sum', 'mean'])

count  sum      mean
                           debt debt      debt
purpose_category                              
операции с автомобилем     4306  403  0.093590
операции с недвижимостью  10811  782  0.072334
получение образования      4013  370  0.092200
проведение свадьбы         2324  186  0.080034

### Вывод 4: Из полученных данных видно, что реже остальных просрачивают возврат кредита заемщики, взявшие кредит на недвижимость - 7,46%, чаще возвращают те, кто брал кредит на проведение свадьбы - 8%, с такой же частотой просрачивают кредит заемщики, взявшие кредит на образование - 8%, на последнем месте стоят те, кто взял кредит на автомобиль - 9,36%.
Заемщики, берущие кредит на недвижимость, более ответственно подходят к выбору суммы, которую они готовы платить ежемесячно, так как кредит берется чаще всего на длительный срок.
Заемщики, берущие кредит на свадьбу, чаще всего возвращают кредит после свадьбы за счет подаренных денег.
Заемщики, берущие кредит на образование, видимо, не всегда оценивают свои возможности (например, работать параллельно на первых курсах технических специальностей очень сложно или практически невозможно) и менее ответственно подходят к своему выбору.
Заемщики, берущие кредит на автомобиль чаще всех просрачивают кредит ввиду того, что чаще ведутся на поводу продавцов в автосалонах и берут автомобиль дороже, чем они могут себе позволить, дополнительно многие не могут оценить сколько будет им впоследствии стоить владение автомобилем (КАСКО, ОСАГО, страхование жизни, обслуживание и содержание автомобиля).

## Общий вывод:

В данном проекте было проанализировано влияние различных параметров на вероятность просрочки кредита клиентом банка для последующего использования данного анализа в модели кредитного скорринга.
Данные были предоставлены заказчиком (банком).
Выводы по данным:
- количество детей: зависимостей выявлено не было, однако клиенты без детей чаще других выплачивают кредит вовремя; недостаточное количество данных про клиентов с 4 и 5 детьми, поэтому данные по этим группам рекомендую не использовать;
- семейное положение: холостые и состоящие в гражанском браке реже других выплачивают креди вовремя;
- уровень дохода: чаще всего кредит выплачивают без просрочек клиенты с не очень высоким уровнем дохода (группа D от 30001 до 50000); недостаточно данных по группам с низким и высоким уровнем дохода, данные по этим группам рекомендую не использовать;
- цели кредита: чаще остальных кредит выплачивают без просрочек клиенты, берущие его на недвижимость.

В данных присутствовали неявные дубликаты из-за того, что использовался разный регистр при написании - рекомендую, чтобы при сборе информации программа автоматически переводила все данные в нижний регистр, независимо от того, как вводилось.
Цели кредита были указаны по-разному - рекомендую в форме, куда вводит свои данные клиент или сотрудник банка, были выведены категории (пользователю так тоже будет удобнее).
У большого числа клиентов был указан не реальный стаж - рекомендую добавить проверку в форму для ввода данных (например если стаж больше 100 лет, то выдовать сообщение о некорректных данных).
Были отрицательные значения в количестве детей и стаже - рекомендую автоматически менять эти значения на положительные или добавить проверку и, в случае выявления отрицательного числа, выдавать сообщение о некорректных данных.
У части клиентов возраст был указан 0 лет - рекомендую добавить проверку при вводе, в случае возраста меньше 18 лет, выдавать сообщение о некорректных данных.
У части клиентов количество детей было указано - 20 - рекомендую добавить проверку и, в случае если количество детей больше 5, то подсветить ячейку ввода, чтобы сотрудник или клент обратили внимание на возможно некорректные данные.